Added a helper function for bc related warp functions

hsalehipour opened the pull request:
Added a helper function for BCs to avoid repeated definition of identical warp functions.

## Contributing Guidelines

<!-- Please make sure you have read and understood our contributing guidelines before submitting this PR -->
- [x] I have read and understood the [CONTRIBUTING.md](https://github.com/Autodesk/XLB/blob/main/CONTRIBUTING.md) guidelines


## Description
We cannot inherit methods in Warp. As a workaround, rather than repeating snippets to represent identical Warp functionalities (which is what is done currently in the code), this PR creates a helper function that keeps those methods. These methods are imported as needed inside construct_warp. 

<!-- 
Thank you for your contribution! Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.
-->


## Type of change

<!-- Please select the options that are relevant. -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so others can reproduce. Include details of your test environment, and the test cases you ran.

- [ ] Test A
- [ ] Test B
-->
- [x] All pytest tests pass

<!-- To run the tests, execute the following command from the root of the repository:

```bash
pytest
```
 -->


## Linting and Code Formatting

Make sure the code follows the project's linting and formatting standards. This project uses **Ruff** for linting.

To run Ruff, execute the following command from the root of the repository:

```bash
ruff check .
```

<!-- You can also fix some linting errors automatically using Ruff:

```bash
ruff check . --fix
```
-->

- [x] Ruff passes

mehdiataei commented:
Please add the helpers to the init and import them from xlb.helper similar to the other helper functions.

hsalehipour commented:
> Please add the helpers to the init and import them from xlb.helper similar to the other helper functions.

It is not possible because the DefaultConfig is not set before xlb.init is called!

mehdiataei commented:
Then this is not ideal. The import of helper functions should not depend on calling xlb.init(). Please find a better solution or consider just consolidating these functions in the BCs.

hsalehipour commented:
Not sure why this is an issue here. This helper function does not call xlb.init() and it cannot be used anywhere ouside of BC implementations where xlb.init() has already been called. Having everything inside __init__ is also not a requirement for packaging. Can you give a scenario where the current implementation might cause an issue?

mehdiataei commented:
A "helper" function should not rely on initialization and must be usable independently. Since these functions may be invoked by users in examples, please relocate them to the boundary condition base class, as their functionality aligns more closely with that context.

hsalehipour commented:
Moving them to the base class is not exteremely messy because they have to be added one by one inside class Init method and given to self. The Warp functions need to be enclosed within a python function to declare the needed constants and variables that are out of the scope of that wp.func! I think the current approach is the cleanest. Please recommend a better approach and I will gladly change it.

hsalehipour commented:
There are few methods right now inside the bc base class that are defined inside the init. I should perhaps move them to this helper function too and import them whenever needed. That would be for this PR. If you agree I can do that as well.

hsalehipour commented:
I moved the new file from helper folder to BC folder which I think addresses the comment about "helper" function.

mehdiataei commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,121 @@
+from typing import Any
+
+# Set the compute and Store dtypes
+if DefaultConfig.default_backend == ComputeBackend.JAX:
```
If DefaultConfig changes or is removed, this code will break!!!

mehdiataei commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,121 @@
+
+# Set local constants
+_d = DefaultConfig.velocity_set.d
+_q = DefaultConfig.velocity_set.q
```
If these variables (like _d, _q, _opp_indices) are global and accessed by other modules, there could be conflicts or unexpected behavior if other parts of the code modify them.

mehdiataei commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,121 @@
+_c = DefaultConfig.velocity_set.c
+_c_float = DefaultConfig.velocity_set.c_float
+_qi = DefaultConfig.velocity_set.qi
+_u_vec = wp.vec(_d, dtype=compute_dtype)
```
Defining constants and objects like wp.vec at the module level means they are created once when the module is loaded. If these objects depend on runtime conditions (e.g., backend or precision), it could lead to incorrect configurations.

mehdiataei reviewed (changes requested):
Please check my comments. The current changes will cause more issues than it tries to solve.

hsalehipour commented:
I agree that BC config should not be strictly identical to DefaultConfig. Based on the previous review I have pushe some changes which I think resolves this issue altogether. Please re-review.

mehdiataei commented on xlb/operator/boundary_condition/bc_regularized.py:
```diff
@@ -127,83 +124,15 @@ def jax_implementation(self, f_pre, f_post, bc_mask, missing_mask):
     def _construct_warp(self):
-        # assign placeholders for both u and rho based on prescribed_value
+        # load helper functions
+        from xlb.operator.boundary_condition.common_warp_functions import CommonWarpFunctions
```
You must import it in the init of the folder instead. _ _ init __

mehdiataei commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,125 @@
+        # Set the default values from the global config
+        self.velocity_set = velocity_set or DefaultConfig.velocity_set
+        self.precision_policy = precision_policy or DefaultConfig.default_precision_policy
+        self.compute_backend = compute_backend or DefaultConfig.default_backend
```
raise error if compute backend is not Warp

mehdiataei commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,125 @@
+            return _f_pre, _f_post, _boundary_id, _missing_mask
+
+        @wp.func
+        def get_bc_fsum(
```
how is this a common function!

hsalehipour replied:
it is common between zouhe and regularized.

mehdiataei commented on xlb/operator/boundary_condition/boundary_condition.py:
```diff
@@ -71,55 +71,26 @@ def __init__(
-
         # Construct some helper warp functions for getting tid data
         if self.compute_backend == ComputeBackend.WARP:
-            self._get_thread_data = _get_thread_data
```
This was a MUCH nicer way...

hsalehipour replied:
Yes we can include all these common methods to the base class but the issue is that not all BCs need those methods and the code becomes less clear in my view with lots of irrelevant functions all inside self.

mehdiataei commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,125 @@
+                        return -_u_vec(_c_float[0, l], _c_float[1, l])
+
+        @wp.func
+        def bounceback_nonequilibrium(
```
This is very BC specific imo. It is common between two very similar BCs only.

hsalehipour replied:
bounceback_nonequilibrium is used in regularized and zouhe and could be used in other BCs. So I would say this is a common function.

mehdiataei commented on xlb/operator/boundary_condition/boundary_condition.py:
```diff
@@ -71,55 +71,26 @@ def __init__(
 
     @partial(jit, static_argnums=(0,), inline=True)
-    def update_bc_auxilary_data(self, f_pre, f_post, bc_mask, missing_mask):
+    def update_bc_auxilary_data_jax(self, f_pre, f_post, bc_mask, missing_mask):
```
What is this change?! You're calling warp functions inside a function called _jax?

We don't use aux data in JAX please remove this.

hsalehipour replied:
This is for the ExtrapolationOutflow BC. And yes we do use that in our JAX implementation.

mehdiataei replied:
But you're not overwriting this here. I don't know how is this fixing any bugs? This is in the base class. If it is not fixing anything there is no point having two abstraction in the base so you can revert it.

mehdiataei replied:
We don't have update_bc_auxilary_data_warp or update_bc_auxilary_data_jax anywhere else.

hsalehipour replied:
Good catch. I pushed new changes and removed them.

mehdiataei reviewed (changes requested):
I’m not a fan of this PR. While I’ve added some comments, this kind of change might seem like an improvement at first glance, but the added abstractions and convolutions unnecessarily complicate what were initially straightforward function calls for someone who is not familiar with this code. Since the number of lines added and removed are nearly identical, it doesn't truly abstract or simplify anything meaningful.

You can just simply revert them back to boundary_condition just as before, and move the function calls related to the BCs to their respective class. For example all functions like bounceback_noneequilibrium, regularize_fpop, _get_fsum are not really "common" functions to be placed in a separate class as they are only applicable to certain BC types. I would rather find them in their respective classes than going to multiple 

I agree get_normal_vectors could be a useful function and can be placed on line 118 of boundary_condition.py.

hsalehipour commented on xlb/operator/boundary_condition/common_warp_functions.py:
```diff
@@ -0,0 +1,125 @@
+from typing import Any
+
+
+class CommonWarpFunctions(object):
```
agreed. I renamed them to be more expressive.

hsalehipour commented:
I addressed all the concerns and pushed some new changes. Although it may seem these changes are cosmetic right now but with new BCs that I have added (under test now) these functions be be used more and it makes more sense (to me) to have them in a central place to help with debugging.

hsalehipour commented:
addressed the latest comment and pushed again.

mehdiataei commented on xlb/operator/boundary_condition/bc_regularized.py:
```diff
@@ -15,9 +15,8 @@
-from xlb.operator.boundary_condition.boundary_condition import ImplementationStep
-from xlb.operator.boundary_condition.boundary_condition_registry import boundary_condition_registry
 from xlb.operator.macroscopic.second_moment import SecondMoment as MomentumFlux
+from xlb.operator.boundary_condition.helper_functions_bc import HelperFunctionsBC
```
Did you add this to _ _ init __ .py file in the boundary condition folder? I don't see this change.

If so, you don't need to import this like this and you can do:

from xlb.operator.boundary_condition import HelperFunctionsBC

hsalehipour replied:
Good point. DONE

mehdiataei commented on xlb/operator/boundary_condition/boundary_condition.py:
```diff
@@ -16,6 +16,7 @@
 from xlb.operator.operator import Operator
 from xlb import DefaultConfig
 from xlb.operator.boundary_condition.boundary_condition_registry import boundary_condition_registry
+from xlb.operator.boundary_condition.helper_functions_bc import HelperFunctionsBC
```
Pls change this

hsalehipour replied:
Done. Pushed it
